src/Microsoft.XmlSerializer.Generator/src/Sgen.cs

```diff
@@ -207,27 +207,49 @@ private void GenerateAssembly(List<string> typeNames, string assemblyName, bool
                 var allMappings = (XmlMapping[])mappings.ToArray(typeof(XmlMapping));
 
                 bool gac = assembly.GlobalAssemblyCache;
-                codePath = codePath == null ? (gac ? Environment.CurrentDirectory : Path.GetDirectoryName(assembly.Location)) : codePath;
+                outputDirectory = outputDirectory == null ? (gac ? Environment.CurrentDirectory : Path.GetDirectoryName(assembly.Location)) : outputDirectory;
                 string serializerName = XmlSerializer.GetXmlSerializerAssemblyName(serializableTypes[0], null);
-                string location = Path.Combine(codePath, serializerName + ".cs");
+                string codePath = Path.Combine(outputDirectory, serializerName + ".cs");
 
                 if (!force)
                 {
-                    if (File.Exists(location))
-                        throw new InvalidOperationException(SR.Format(SR.ErrSerializerExists, location, "force"));
+                    if (File.Exists(codePath))
+                        throw new InvalidOperationException(SR.Format(SR.ErrSerializerExists, codePath, "force"));
                 }
 
-                if (Directory.Exists(location))
+                if (Directory.Exists(codePath))
                 {
-                    throw new InvalidOperationException(SR.Format(SR.ErrDirectoryExists, location));
+                    throw new InvalidOperationException(SR.Format(SR.ErrDirectoryExists, codePath));
                 }
 
-                bool success = XmlSerializer.GenerateSerializer(serializableTypes, allMappings, codePath);
+                if (!Directory.Exists(outputDirectory))
+                {
+                    throw new ArgumentException(SR.Format(SR.ErrDirectoryNotExists, codePath, outputDirectory));
+                }
+
+                bool success;
+
+                try
+                {
+                    if (File.Exists(codePath))
+                    {
+                        File.Delete(codePath);
+                    }
+
+                    using (FileStream fs = File.Create(codePath))
+                    {
+                        success = XmlSerializer.GenerateSerializer(serializableTypes, allMappings, fs);
+                    }
+                }
+                catch (UnauthorizedAccessException)
```

Comment:
It will throw IOException on the earlier version of OS like XP. I tested on my machine and the text file will be deleted although I open it.